### ISSUE TITLE
LWF implementation is different than the origianl paper.

### ISSUE THREAD

**Issue by yonatansverdlov** [ContinualAI/avalanche#1223]
I know it was asked, but the original implementation uses the inner product between the current probabilities vector and the previous log previous probabilities vector and your one uses KL loss.
For example, if Pcurr = Pprev then the KL loss is zero but the first loss is non-zero.
It shows that those losses are different.
In addition, the implementation in https://github.com/MasLiang/Learning-without-Forgetting-using-Pytorch/blob/main/main.py/line 161 uses the inner product method.
One possible solution is to have two versions of the same method.
Thanks!

**Comment by AntonioCarta** [comment]
Can you write down the equations? I think our implementation corresponds to the original LwF [1](https://arxiv.org/abs/1606.09282). We use the `kl_div` method but the resulting equations should those of [1].

This is our distillation loss:
```
    def _distillation_loss(self, out, prev_out, active_units):
        """Compute distillation loss between output of the current model and
        and output of the previous (saved) model.
        """
        # we compute the loss only on the previously active units.
        au = list(active_units)

        log_p = torch.log_softmax(out[:, au] / self.temperature, dim=1)
        q = torch.softmax(prev_out[:, au] / self.temperature, dim=1)
        res = torch.nn.functional.kl_div(log_p, q, reduction="batchmean")
        return res
```

**Comment by yonatansverdlov** [comment]
The equation there is: 
L(p_curr,p_prev) = sum( p_curr[i]*-log(p_prev[i])) = <p_curr, -log(p_prev)> as in equation 1 in the paper.
In your case KL divergence does the following:
p_prev *(log(p_prev)-log(p_curr)) = p_prev * log(p_prev / p_curr).
Now, the expressions are not the same as we can see.
I think your option is better but this should be pointed out.

**Comment by AntonioCarta** [comment]
we are doing `kl_div(log_p_curr, p_prev) = p_prev * (log (p_prev / p_curr)) = p_prev * log(p_prev) - p_prev * log(p_curr)`. 
Now, the first term is constant (we don't optimize the teacher), so optimizing the crossentropy and kl-div are equivalent.

**Comment by yonatansverdlov** [comment]
You are right!
Pointing this out in my opinion is important.
Thanks!

**Comment by AntonioCarta** [comment]
fixed in #1237
